plugins/vbe/fu-vbe-simple-device.c

```diff
@@ -0,0 +1,844 @@
+	       GError **error)
+{
+	gint count, i;
+
+	count = fit_cfg_img_count(fit, cfg, "firmware");
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+
+	for (i = 0; i < count; i++) {
+		gint image = fit_cfg_img(fit, cfg, "firmware", i);
+
+		if (image < 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_READ,
+				    "'firmware' image #%d has no node",
+				    i);
+			return FALSE;
+		}
+		fu_progress_set_percentage_full(progress, i, count);
+
+		if (!process_image(self, fit, image, error))
+			return FALSE;
+	}
+	fu_progress_set_percentage_full(progress, i, count);
+	fu_progress_step_done(progress);
+
+	return TRUE;
+}
+
+/**
+ * process_fit() - Write the firmware from a FIT to the device
+ *
+ * Select the best configuration for the model, then write that configuration
+ * to the device.
+ *
+ * @fit: FIT to write
+ * @self: Device to use
+ * @compat_list: List of compatible properties for this model, if any
+ * @progress: Progress information to update (this is simplistic and assumes
+ * that each image takes the same time to write)
+ * @error: Returns error information if this function returns FALSE
+ * Returns: TRUE on success, FALSE on failure
+ */
+static gboolean
+process_fit(FuVbeSimpleDevice *self,
+	    struct fit_info *fit,
+	    GList *compat_list,
+	    FuProgress *progress,
+	    GError **error)
+{
+	struct vbe_simple_state *state = &self->state;
+	gint best_prio = INT_MAX;
+	const gchar *cfg_name;
+	const gchar *version;
+	const GList *entry;
+	gint cfg_count = 0;
+	gint best_cfg = 0;
+	gint cfg;
+
+	g_debug("model: ");
+	for (entry = g_list_first(compat_list); entry; entry = g_list_next(entry))
+		g_debug("   %s", (gchar *)entry->data);
+
+	for (cfg = fit_first_cfg(fit); cfg > 0; cfg_count++, cfg = fit_next_cfg(fit, cfg)) {
+		gint prio = check_config_match(fit, cfg, compat_list);
+		g_debug("config '%s': priority=%d", fit_cfg_name(fit, cfg), prio);
+		if (prio >= 0 && (!best_cfg || prio < best_prio)) {
+			best_cfg = cfg;
+			best_prio = prio;
+		}
+	}
+
+	g_debug("cfg_count=%d, best_cfg=%d", cfg_count, best_cfg);
+	if (!cfg_count) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_READ, "FIT has no configurations");
+		return FALSE;
+	}
+
+	if (!best_cfg) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_READ, "No matching configuration");
+		return FALSE;
+	}
+	cfg_name = fit_cfg_name(fit, best_cfg);
+	version = fit_cfg_version(fit, best_cfg);
+	if (!version) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "Configuration '%s' has no version",
+			    cfg_name);
+		return FALSE;
+	}
+
+	g_debug("Best configuration: '%s', priority %d, version %s", cfg_name, best_prio, version);
+
+	if (!process_config(self, fit, best_cfg, progress, error))
+		return FALSE;
+
+	g_free(state->cur_version);
+	g_free(state->status);
+	state->finish_time = time(NULL);
+	state->cur_version = g_strdup(version);
```

Comment:
@sjg20 I'm a bit confused why we're saving the version to a separate file here; can't it be read out from `cur-version` during the next-reboot coldplug?

Reply:
> Using the filesystem is not very helpful

Yup, I really think we don't want to do the /var/lib/fwupd/vbe/simple.dtb thing -- it's storing the data in the wrong "layer". I like your `nvdata` proposal -- i.e. the version has to be read from somewhere in the flash image, not from a partition.

Reply:
Well nvdata is for firmware since it cannot always access simple.dtb - having said that, fwupd will eventually have to be able to update nvdata.

My thinking is that for user space / fwupd we need richer information about what needs doing and what has already been tried, e.g. to handle rollbacks and limits on how far to roll forward. We should be able to signal what needs doing from firmware, but fwupd needs to actually do it. Quite apart from that we need debugging and logging which is tricky to do across reboots, so recording recent updates and why VBE did what it did is helpful.

Let's say (in an A/B situation) we select an update and boot it, only to find that it does not boot into user space. We reboot and try the old firmware, which works. We now need to go back to the old firmware in both slots (A and B).

The problem with vbe-simple is that it oes not have to deal with these cases. On the other hand if we don't have the feature included, we are not defining how to do it with more complex VBE methods.

Still. I don't have a clear idea of exactly what is needed in this file and vbe-simple does not need it, so I am OK with dropping it for now.